### ISSUE TITLE
Passing parameters to sql query.

### ISSUE THREAD

**Issue by trara538** [Vanuatu-National-Statistics-Office/vnso-RAP-tradeStats-materials#78]
I need some help on how to pass a parameter from a SelectInput value to an SQL query for purpose of retrieving specific records from the database.

**Comment by JosephCrispell** [comment]
Hi @trara538,

Sorry it took me so long to get to this! I've made a very minor change to your code that I think will allow you to feed a user selection into a MySQL query:
https://github.com/Vanuatu-National-Statistics-Office/vnso-RAP-tradeStats-materials/blob/92935b72d4b46a08532df30688770404d9608030/R/dashboard.Rmd#L465-L472

Let me know if the above change works - difficult for me to test on my end because I don't have the same local MySQL set up.

**Comment by JosephCrispell** [comment]
Hi @trara538,

I've got a working versions of passing shiny input parameters into a MySQL and generating some output using a shiny app and using flexdashboard 🎉:

- flexdashboard: [RMySQL_flexdashboard.Rmd](https://github.com/Vanuatu-National-Statistics-Office/vnso-RAP-tradeStats-materials/blob/Create_Flexdashboard/R/RMySQL-shiny_examples/RMySQL_flexdashboard.Rmd)
- Shiny application: [RMySQL_shiny-app.R](https://github.com/Vanuatu-National-Statistics-Office/vnso-RAP-tradeStats-materials/blob/Create_Flexdashboard/R/RMySQL-shiny_examples/RMySQL_shiny-app.R)

See if you can get the above examples to work for you (they are calling an MySQL out-of-the-box table so will hopefully work). Then see if you can modify to work in dashboard. I think part of the problem is that the query was working but the outputs were being hidden. Let me know how you get on.

**Comment by JosephCrispell** [comment]
Hi @trara538,

I think that's the demo's working now 🤞. The issue was caused by a connection issue as we thought (more information on it [here](https://stackoverflow.com/questions/33885400/rmysql-internal-error-in-rs-dbi-getconnection-corrupt-connection-handle)) and the solution was to connect and disconnect from the database for each query and I've written a simple set of  functions to do this for us:
https://github.com/Vanuatu-National-Statistics-Office/vnso-RAP-tradeStats-materials/blob/868afdb1c4abc32bc4478a317c01a0b5ff823299/R/RMySQL-shiny_examples/RMySQL_flexdashboard.Rmd#L24-L55

I've also switched the password to read in from an [environmental variable](https://support.shotgunsoftware.com/hc/en-us/articles/114094235653-Setting-global-environment-variables-on-Windows#:~:text=Windows%207,to%20edit%2C%20and%20click%20Edit.) (as it's more secure). Note that I had to store the password in a variable rather than parameter because from some reason the `reactive({})` element doesn't recognise the `params` variable (!?!).

Anyways both demos are now working well - give them a try on your computer and let me know how you get on. Almost there! 😊

**Comment by trara538** [comment]
Hi Joe,

Just a little bit confused on how you are using the password viriable here intstead of using the password text instead. You mentioned something about concealing your password which is a great idea. I just could not figure out how to do this.

**Comment by JosephCrispell** [comment]
Hi @trara538,

Firstly, if you would like to use a password as a string you still can just replace `Sys.getenv("MYSQLPASS")` at line 19 in the [`RMySQL_flexdashboard.Rmd`](https://github.com/Vanuatu-National-Statistics-Office/vnso-RAP-tradeStats-materials/blob/Create_Flexdashboard/R/RMySQL-shiny_examples/RMySQL_flexdashboard.Rmd#L19) file with your password as a string (for example `sqlPassword <- "passwordHere"`).

https://github.com/Vanuatu-National-Statistics-Office/vnso-RAP-tradeStats-materials/blob/868afdb1c4abc32bc4478a317c01a0b5ff823299/R/RMySQL-shiny_examples/RMySQL_flexdashboard.Rmd#L19 

The existing approach uses an environmental variable. An environmental variable is a variable that is created on your computer (usually outside R) that you can access and use in the command line, via a program, or in a programming language (better explanation [here](https://medium.com/chingu/an-introduction-to-environment-variables-and-how-to-use-them-f602f66d15fa)). What I've done is created an environmental variable called `MYSQLPASS` and it's value is my password (to create a environmental variable follow these [instructions](https://superuser.com/questions/284342/what-are-path-and-other-environment-variables-and-how-can-i-set-or-use-them)). I then use the `Sys.getenv()` function in R to return the value (my password as a character string) of my environmental variable by passing the name of my environmental variable into the function (`Sys.getenv("MYSQLPASS")`). The advantage of this approach is that my password is never entered into my code where others may see it, and it is also accessible to other scripts I develop.

To summarise, to get the code working on your computer (using an environmental variable) follow these steps:

1. Create an environmental variable called `MYSQLPASS` and assign it's value to be your password ([instructions](https://superuser.com/questions/284342/what-are-path-and-other-environment-variables-and-how-can-i-set-or-use-them))
2. Run the [`RMySQL_flexdashboard.Rmd`](https://github.com/Vanuatu-National-Statistics-Office/vnso-RAP-tradeStats-materials/blob/Create_Flexdashboard/R/RMySQL-shiny_examples/RMySQL_flexdashboard.Rmd#L19) to start the shiny application

*Note: the environmental variable can be called anything (all uppercase is the standard format for names)*

**Comment by JosephCrispell** [comment]
Hi @trara538, 

Following our meeting today, I've updated the  [RMySQL_flexdashboard.Rmd](https://github.com/Vanuatu-National-Statistics-Office/vnso-RAP-tradeStats-materials/blob/868afdb1c4abc32bc4478a317c01a0b5ff823299/R/RMySQL-shiny_examples/RMySQL_flexdashboard.Rmd#L19) demo. 

With the demo it should now be possible to:
- Set the username in first cell (remember if you can't get environmental variable to work just replace the `Sys.getenv("MYSQLPASS")` with your password in a string
- Select a text based value and feed that into a MySQL query

Remember that I'm having issues with the output showing up in my browser and finding that I need to open the application in the browser, make a minor change to the R code and save it, then refresh page in browser in order for the output table to appear.

Let me know how you get on.
